aya/src/maps/xdp/cpu_map.rs

```diff
@@ -0,0 +1,185 @@
+            .ok_or(MapError::KeyNotFound)
+    }
+
+    /// An iterator over the elements of the map.
+    pub fn iter(&self) -> impl Iterator<Item = Result<CpuMapValue, MapError>> + '_ {
+        (0..self.len()).map(move |i| self.get(i, 0))
+    }
+}
+
+impl<T: BorrowMut<MapData>> CpuMap<T> {
+    /// Sets the queue size at the given CPU index, and optionally a chained program.
+    ///
+    /// When sending the packet to the CPU at the given index, the kernel will queue up to
+    /// `queue_size` packets before dropping them.
+    ///
+    /// Another XDP program can be passed in that will be run on the target CPU, instead of the CPU
+    /// that receives the packets. This allows to perform minimal computations on CPUs that
+    /// directly handle packets from a NIC's RX queues, and perform possibly heavier ones in other,
+    /// less busy CPUs.
+    ///
+    /// Note that only XDP programs with the `map = "cpumap"` argument can be passed. See the
+    /// kernel-space `aya_bpf::xdp` for more information.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
+    /// if `bpf_map_update_elem` fails, [`MapError::ProgIdNotSupported`] if the kernel does not
+    /// support program ids and one is provided.
+    pub fn set(
+        &mut self,
+        cpu_index: u32,
+        queue_size: u32,
+        program: Option<&ProgramFd>,
+        flags: u64,
+    ) -> Result<(), MapError> {
+        let data = self.inner.borrow_mut();
+        check_bounds(data, cpu_index)?;
+        let fd = data.fd;
+
+        let res = if FEATURES.cpumap_prog_id() {
+            let value = bpf_cpumap_val {
+                qsize: queue_size,
+                bpf_prog: bpf_cpumap_val__bindgen_ty_1 {
+                    // Default is valid as the kernel will only consider fd > 0:
+                    // https://github.com/torvalds/linux/blob/2dde18cd1d8fac735875f2e4987f11817cc0bc2c/kernel/bpf/cpumap.c#L466
+                    fd: program
+                        .map(|prog| prog.as_fd().as_raw_fd())
+                        .unwrap_or_default(),
+                },
+            };
+            bpf_map_update_elem(fd, Some(&cpu_index), &value, flags)
+        } else {
+            if program.is_some() {
+                return Err(MapError::ProgIdNotSupported);
+            }
+            bpf_map_update_elem(fd, Some(&cpu_index), &queue_size, flags)
+        };
+
+        res.map_err(|(_, io_error)| SyscallError {
+            call: "bpf_map_update_elem",
+            io_error,
+        })?;
+        Ok(())
+    }
+}
+
+impl<T: Borrow<MapData>> IterableMap<u32, CpuMapValue> for CpuMap<T> {
+    fn map(&self) -> &MapData {
+        self.inner.borrow()
+    }
+
+    fn get(&self, key: &u32) -> Result<CpuMapValue, MapError> {
+        self.get(*key, 0)
+    }
+}
+
+unsafe impl Pod for bpf_cpumap_val {}
+
+#[derive(Clone, Copy, Debug)]
+pub struct CpuMapValue {
+    pub qsize: u32,
```

Comment:
This should probably be queue_size? As a rule of thumb let's not use kernel
names, ever :P

Reply:
gotcha!